File: ZSDK_DevDemos_Combined/DeveloperDemosVisualBasic/StoredFormatDemoForm.vb
' ********************************************** 
' * CONFIDENTIAL AND PROPRIETARY 
' *
' * The source code and other information contained herein is the confidential and the exclusive property of
' * ZIH Corp. and is subject to the terms and conditions in your end user license agreement.
' * This source code, and any other information contained herein, shall not be copied, reproduced, published, 
' * displayed or distributed, in whole or in part, in any medium, by any means, for any purpose except as
' * expressly permitted under such license agreement.
' * 
' * Copyright ZIH Corp. 2010
' *
' * ALL RIGHTS RESERVED 
' *********************************************** 

Imports System
Imports System.Collections.Generic
Imports System.ComponentModel
Imports System.Drawing
Imports System.Threading
Imports System.Windows.Forms
Imports ZSDK_API.ApiException
Imports ZSDK_API.Comm
Imports ZSDK_API.Printer
Imports System.Text

Namespace ZebraDeveloperDemos
    <System.ComponentModel.DesignerCategory("Code")> _
    Partial Public Class StoredFormatDemoForm
        Inherits ZebraForm

        Private labels As New List(Of Label)()
        Private userInputVars As New List(Of TextBox)()

        Private Delegate Sub GenericEventHandler()

        Public Sub New()
            InitializeComponent()
            connectionPanel.BringToFront()
            zebraConnectionPanel = connectionPanel
            AddHandler connectionEstablished, (AddressOf Connected)
            AddHandler connectionClosed, (AddressOf Disconnected)
        End Sub

        Private Sub leftSideButtonPressed(ByVal sender As Object, ByVal e As EventArgs)
            If leftSideButton.Text.Equals("Back") Then
                leftSideButton.Enabled = False
                RemoveVariableFields()
                formatListPanel.BringToFront()
                Invoke(New GenericEventHandler(AddressOf ConnectedButtonSettings))
            ElseIf leftSideButton.Text.Equals("Disconnect") Then
                toggleMenuButtons(False)
                connectionPanel.BringToFront()
                disconnect()
            End If
        End Sub

        Private Sub RemoveVariableFields()
            SuspendLayout()
            For i As Integer = 0 To labels.Count - 1
                varFieldPanel.Controls.Remove(labels(i))
                labels(i).Dispose()
                varFieldPanel.Controls.Remove(userInputVars(i))
                userInputVars(i).Dispose()
            Next
            quantity.Value = 1
            labels.Clear()
            labels.TrimExcess()
            userInputVars.Clear()
            userInputVars.TrimExcess()
            varFieldPanel.AutoScrollPosition = New Point(0, 0)
            ResumeLayout(False)
        End Sub

        Private Sub Connected()
            Invoke(New GenericEventHandler(AddressOf ConnectedButtonSettings))
            Invoke(New GenericEventHandler(AddressOf DisplayFormatsThread))
        End Sub

        Private Sub ConnectedButtonSettings()
            formatListPanel.BringToFront()
            leftSideButton.Text = "Disconnect"
            rightSideButton.Text = "Refresh"
            toggleMenuButtons(True)
        End Sub

        Private Sub Disconnected()
            Invoke(New GenericEventHandler(AddressOf NoConnectionButtonSettings))
        End Sub

        Private Sub NoConnectionButtonSettings()
            rightSideButton.Text = "List Formats"
            rightSideButton.Enabled = True
            leftSideButton.Text = ""
            leftSideButton.Enabled = False
        End Sub


        Private Sub DisplayFormatsThread()
            formatListPanel.BringToFront()
            Dim t As New Thread(AddressOf doDisplayFormatsThread)
            t.Start()
        End Sub

        Private Sub doDisplayFormatsThread()
            Try
                GetListOfFormats()
            Catch generatedExceptionName As ZebraException
                Invoke(New GenericEventHandler(AddressOf bringConnectionPanelToFront))
                updateGuiFromWorkerThread("COMM Error! Disconnected", Color.Red)
            End Try
        End Sub

        Private Sub bringConnectionPanelToFront()
            connectionPanel.BringToFront()
        End Sub

        Private Delegate Sub FormatListEventHandler(ByVal formatList As List(Of [String]))

        Private Sub GetListOfFormats()
            Dim connection As ZebraPrinterConnection = getConnection()
            Dim printer As ZebraPrinter = ZebraPrinterFactory.GetInstance(connection)

            Dim zplFormatExtensions As [String]() = New [String]() {"ZPL"}
            Dim cpclFormatExtensions As [String]() = New [String]() {"FMT", "LBL"}
            Dim extensions As [String]() = If(printer.GetPrinterControlLanguage() = PrinterLanguage.ZPL, zplFormatExtensions, cpclFormatExtensions)

            Dim formatList As New List(Of [String])()
            For Each format As [String] In printer.GetFileUtil().RetrieveFileNames(extensions)
                formatList.Add(format)
            Next
            Invoke(New FormatListEventHandler(AddressOf DisplayNames), New Object() {formatList})
        End Sub

        Private Sub DisplayNames(ByVal formatList As List(Of [String]))
            formatListBox.DataSource = formatList
            formatListBox.Refresh()
            formatSelectButton.Enabled = True
            toggleMenuButtons(True)
        End Sub

        Private Sub rightSideButtonPressed(ByVal sender As Object, ByVal e As EventArgs)
            toggleMenuButtons(False)
            If rightSideButton.Text.Equals("List Formats") Then
                connect()
            ElseIf rightSideButton.Text.Equals("Print") Then
                SendLabelsToPrinter()
                toggleMenuButtons(True)
            ElseIf rightSideButton.Text.Equals("Refresh") Then
                formatSelectButton.Enabled = False
                formatListBox.DataSource = Nothing
                Invoke(New GenericEventHandler(AddressOf DisplayFormatsThread))
            End If
        End Sub

        Private Sub SendLabelsToPrinter()
            Dim varFieldNames As FieldDescriptionData() = GetVarFieldNames()
            Dim fieldNumbersAndNames As New Dictionary(Of Integer, [String])()
            For i As Integer = 0 To labels.Count - 1
                Dim fnNumber As Integer = varFieldNames(i).FieldNumber
                fieldNumbersAndNames(fnNumber) = userInputVars(i).Text
            Next
            PrintLabels(formatListBox.SelectedItem.ToString(), fieldNumbersAndNames, Convert.ToInt32(quantity.Value))
            printJobSentPanel.BringToFront()
        End Sub

        Private formatName As [String]
        Private fieldNumbersAndNames As Dictionary(Of Integer, [String])
        Private numLabels As Integer

        Private Sub PrintLabels(ByVal formatName As [String], ByVal fieldNumbersAndNames As Dictionary(Of Integer, [String]), ByVal qty As Integer)
            Me.formatName = formatName
            Me.fieldNumbersAndNames = fieldNumbersAndNames
            Me.numLabels = qty

            Dim t As New Thread(AddressOf doPrintLabels)
            t.Start()
        End Sub

        Private Sub doPrintLabels()
            For i As Integer = 0 To Me.numLabels - 1
                getPrinter().GetFormatUtil().PrintStoredFormat(Me.formatName, Me.fieldNumbersAndNames, "utf-8")
            Next
        End Sub

        Private Sub formatSelectButton_Click(ByVal sender As Object, ByVal e As EventArgs)
            If formatListBox.SelectedItem IsNot Nothing Then
                RemoveVariableFields()
                SuspendLayout()
                Invoke(New GenericEventHandler(AddressOf BackAndPrintButtonUpdate))
                Dim varFieldNames As FieldDescriptionData() = GetVarFieldNames()

                titleLabel.Text = "Format " & formatListBox.SelectedItem.ToString() & ":"

                For i As Integer = 0 To varFieldNames.Length - 1
                    Dim labelText As [String] = If(varFieldNames(i).FieldName Is Nothing, Convert.ToString(varFieldNames(i).FieldNumber), varFieldNames(i).FieldName)
                    Dim label As New Label()
                    label.Text = labelText & " :"
                    labels.Add(label)
                    userInputVars.Add(New TextBox())
                Next

                layoutVariables()
                ResumeLayout(False)
            End If
        End Sub

        Private Sub layoutVariables()
            Const HorizontalPositionOfLabelStart As Integer = 21
            Const VerticalPositionOfLabelStart As Integer = 55
            Const SpacingBetweenGuiElements As Integer = 30
            For i As Integer = 0 To labels.Count - 1
                labels(i).Location = New Point(HorizontalPositionOfLabelStart, VerticalPositionOfLabelStart + (i * SpacingBetweenGuiElements))
                labels(i).Size = New Size(65, 20)
                userInputVars(i).Location = New Point(HorizontalPositionOfLabelStart + 80, VerticalPositionOfLabelStart + (i * SpacingBetweenGuiElements))
                userInputVars(i).Size = New Size(120, 20)
            Next

            Dim verticalLoc As Integer = VerticalPositionOfLabelStart + (labels.Count * SpacingBetweenGuiElements)
            quantityLabel.Location = New Point(HorizontalPositionOfLabelStart, verticalLoc + 5)
            quantity.Location = New Point(HorizontalPositionOfLabelStart + 110, verticalLoc)

            AutoScaleDimensions = New SizeF(96.0F, 96.0F)

            For i As Integer = 0 To labels.Count - 1
                varFieldPanel.Controls.Add(labels(i))
                varFieldPanel.Controls.Add(userInputVars(i))
            Next
        End Sub

        Private Sub BackAndPrintButtonUpdate()
            varFieldPanel.BringToFront()
            leftSideButton.Text = "Back"
            rightSideButton.Text = "Print"
        End Sub

        Private Function GetVarFieldNames() As FieldDescriptionData()
            Dim formatName As [String] = formatListBox.SelectedItem.ToString()
            Dim fieldNames As FieldDescriptionData() = New FieldDescriptionData() {}
            Try
                Dim printer As ZebraPrinter = getPrinter()
                Dim formatData As Byte() = printer.GetFormatUtil().RetrieveFormatFromPrinter(formatName)
                Dim formatString As [String] = Encoding.UTF8.GetString(formatData, 0, formatData.Length)
                fieldNames = printer.GetFormatUtil().GetVariableFields(formatString)
            Catch generatedExceptionName As ZebraPrinterConnectionException
                updateGuiFromWorkerThread("Communication Error", Color.Red)
            End Try
            Return fieldNames
        End Function

        Private Sub toggleMenuButtons(ByVal state As Boolean)
            rightSideButton.Enabled = state
            leftSideButton.Enabled = state

        End Sub

        Private Sub reprintButton_Click(ByVal sender As Object, ByVal e As EventArgs)
            SendLabelsToPrinter()
        End Sub

        Private Sub exitButton_Click(ByVal sender As Object, ByVal e As EventArgs)
            Close()
        End Sub

        Private Sub backToListFormatsButton_Click(ByVal sender As Object, ByVal e As EventArgs)
            ConnectedButtonSettings()
        End Sub

        Private Sub reEnterDataButton_Click(ByVal sender As Object, ByVal e As EventArgs)
            varFieldPanel.BringToFront()
        End Sub

        Private Sub StoredFormatDemoForm_Closing(ByVal sender As Object, ByVal e As CancelEventArgs)
            If getConnection() IsNot Nothing Then
                getConnection().Close()
            End If
            RemoveVariableFields()
        End Sub

    End Class
End Namespace

File: ZSDK_DevDemos_Combined/DeveloperDemosCSharp/StoredFormatDemoForm.cs
/********************************************** 
 * CONFIDENTIAL AND PROPRIETARY 
 *
 * The source code and other information contained herein is the confidential and the exclusive property of
 * ZIH Corp. and is subject to the terms and conditions in your end user license agreement.
 * This source code, and any other information contained herein, shall not be copied, reproduced, published, 
 * displayed or distributed, in whole or in part, in any medium, by any means, for any purpose except as
 * expressly permitted under such license agreement.
 * 
 * Copyright ZIH Corp. 2010
 *
 * ALL RIGHTS RESERVED 
 ***********************************************/

using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Drawing;
using System.Threading;
using System.Windows.Forms;
using ZSDK_API.ApiException;
using ZSDK_API.Comm;
using ZSDK_API.Printer;
using System.Text;

namespace ZebraDeveloperDemos {

    public partial class StoredFormatDemoForm : ZebraForm {

        List<Label> labels = new List<Label>();
        List<TextBox> userInputVars = new List<TextBox>();

        private delegate void GenericEventHandler();

        public StoredFormatDemoForm() {
            InitializeComponent();
            connectionPanel.BringToFront();
            zebraConnectionPanel = connectionPanel;
            connectionEstablished += (Connected);
            connectionClosed += (Disconnected);
        }

        private void leftSideButtonPressed(object sender, EventArgs e) {
            if (leftSideButton.Text.Equals("Back")) {
                leftSideButton.Enabled = false;
                RemoveVariableFields();
                formatListPanel.BringToFront();
                Invoke(new GenericEventHandler(ConnectedButtonSettings));
            } else if (leftSideButton.Text.Equals("Disconnect")) {
                toggleMenuButtons(false);
                connectionPanel.BringToFront();
                disconnect();
            }
        }

        private void RemoveVariableFields() {
            SuspendLayout();
            for (int i = 0; i < labels.Count; i++) {
                varFieldPanel.Controls.Remove(labels[i]);
                labels[i].Dispose();
                varFieldPanel.Controls.Remove(userInputVars[i]);
                userInputVars[i].Dispose();
            }
            quantity.Value = 1;
            labels.Clear();
            labels.TrimExcess();
            userInputVars.Clear();
            userInputVars.TrimExcess();
            varFieldPanel.AutoScrollPosition = new Point(0,0);
            ResumeLayout(false);
        }

        private void Connected() {
            Invoke(new GenericEventHandler(ConnectedButtonSettings));
            Invoke(new GenericEventHandler(DisplayFormatsThread));
        }

        private void ConnectedButtonSettings() {
            formatListPanel.BringToFront();
            leftSideButton.Text = "Disconnect";
            rightSideButton.Text = "Refresh";
            toggleMenuButtons(true);
        }

        private void Disconnected() {
            Invoke(new GenericEventHandler(NoConnectionButtonSettings));
        }

        private void NoConnectionButtonSettings() {
            rightSideButton.Text = "List Formats";
            rightSideButton.Enabled = true;
            leftSideButton.Text = "";
            leftSideButton.Enabled = false;
        }


        private void DisplayFormatsThread() {
            formatListPanel.BringToFront();
            Thread t = new Thread(doDisplayFormatsThread);
            t.Start();
        }

        private void doDisplayFormatsThread() {
            try {
                GetListOfFormats();
            } catch (ZebraException) {
                Invoke(new GenericEventHandler(bringConnectionPanelToFront));
                updateGuiFromWorkerThread("COMM Error! Disconnected", Color.Red);
            }
        }

        private void bringConnectionPanelToFront() {
            connectionPanel.BringToFront();
        }

        private delegate void FormatListEventHandler(List<String> formatList);

        private void GetListOfFormats() {
            ZebraPrinterConnection connection = getConnection();
            ZebraPrinter printer = ZebraPrinterFactory.GetInstance(connection);

            String[] zplFormatExtensions = new String[] { "ZPL" };
            String[] cpclFormatExtensions = new String[] { "FMT", "LBL" };
            String[] extensions = printer.GetPrinterControlLanguage() == PrinterLanguage.ZPL ? zplFormatExtensions : cpclFormatExtensions;

            List<String> formatList = new List<String>();
            foreach (String format in printer.GetFileUtil().RetrieveFileNames(extensions)) {
                formatList.Add(format);
            }
            Invoke(new FormatListEventHandler(DisplayNames), new object[] { formatList });
        }

        private void DisplayNames(List<String> formatList) {
            formatListBox.DataSource = formatList;
            formatListBox.Refresh();
            formatSelectButton.Enabled = true;
            toggleMenuButtons(true);
        }

        private void rightSideButtonPressed(object sender, EventArgs e) {
            toggleMenuButtons(false);
            if (rightSideButton.Text.Equals("List Formats")) {
                connect();
            } else if (rightSideButton.Text.Equals("Print")) {
                SendLabelsToPrinter();
                toggleMenuButtons(true);
            } else if (rightSideButton.Text.Equals("Refresh")) {
                formatSelectButton.Enabled = false;
                formatListBox.DataSource = null;
                Invoke(new GenericEventHandler(DisplayFormatsThread));
            }
        }

        private void SendLabelsToPrinter() {
            FieldDescriptionData[] varFieldNames = GetVarFieldNames();
            Dictionary<int, String> fieldNumbersAndNames = new Dictionary<int, String>();
            for (int i = 0; i < labels.Count; i++) {
                int fnNumber = varFieldNames[i].FieldNumber;
                fieldNumbersAndNames[fnNumber] = userInputVars[i].Text;
            }
            PrintLabels(formatListBox.SelectedItem.ToString(), fieldNumbersAndNames, Convert.ToInt32(quantity.Value));
            printJobSentPanel.BringToFront();
        }

        private String formatName;
        private Dictionary<int, String> fieldNumbersAndNames;
        private int numLabels;

        private void PrintLabels(String formatName, Dictionary<int, String> fieldNumbersAndNames, int qty) {
            this.formatName = formatName;
            this.fieldNumbersAndNames = fieldNumbersAndNames;
            this.numLabels = qty;

            Thread t = new Thread(doPrintLabels);
            t.Start();
        }

        private void doPrintLabels() {
            for (int i = 0; i < this.numLabels; i++) {
                getPrinter().GetFormatUtil().PrintStoredFormat(this.formatName, this.fieldNumbersAndNames, "utf-8");
            }
        }

        private void formatSelectButton_Click(object sender, EventArgs e) {
            if (formatListBox.SelectedItem != null) {
                RemoveVariableFields();
                SuspendLayout();
                Invoke(new GenericEventHandler(BackAndPrintButtonUpdate));
                FieldDescriptionData[] varFieldNames = GetVarFieldNames();

                titleLabel.Text = "Format " + formatListBox.SelectedItem.ToString() + ":";

                for (int i = 0; i < varFieldNames.Length; i++)
                {
                    String labelText = varFieldNames[i].FieldName == null ? Convert.ToString(varFieldNames[i].FieldNumber) : varFieldNames[i].FieldName;
                    Label label = new Label();
                    label.Text = labelText + " :";
                    labels.Add(label);
                    userInputVars.Add(new TextBox());
                }

                layoutVariables();
                ResumeLayout(false);
            }
        }

        private void layoutVariables() {
            const int HorizontalPositionOfLabelStart = 21;
            const int VerticalPositionOfLabelStart = 55;
            const int SpacingBetweenGuiElements = 30;
            for (int i = 0; i < labels.Count; i++) {
                labels[i].Location = new Point(HorizontalPositionOfLabelStart, VerticalPositionOfLabelStart + (i * SpacingBetweenGuiElements));
                labels[i].Size = new Size(65, 20);
                userInputVars[i].Location = new Point(HorizontalPositionOfLabelStart + 80, VerticalPositionOfLabelStart + (i * SpacingBetweenGuiElements));
                userInputVars[i].Size = new Size(120, 20);
            }

            int verticalLoc = VerticalPositionOfLabelStart + (labels.Count * SpacingBetweenGuiElements);
            quantityLabel.Location = new Point(HorizontalPositionOfLabelStart, verticalLoc + 5);
            quantity.Location = new Point(HorizontalPositionOfLabelStart + 110, verticalLoc);

            AutoScaleDimensions = new SizeF(96F, 96F);

            for (int i = 0; i < labels.Count; i++) {
                varFieldPanel.Controls.Add(labels[i]);
                varFieldPanel.Controls.Add(userInputVars[i]);
            }
        }

        private void BackAndPrintButtonUpdate() {
            varFieldPanel.BringToFront();
            leftSideButton.Text = "Back";
            rightSideButton.Text = "Print";
        }

        private FieldDescriptionData[] GetVarFieldNames() {
            String formatName = formatListBox.SelectedItem.ToString();
            FieldDescriptionData[] fieldNames = new FieldDescriptionData[] { };
            try {
                ZebraPrinter printer = getPrinter();
                byte[] formatData = printer.GetFormatUtil().RetrieveFormatFromPrinter(formatName);
                String formatString = Encoding.UTF8.GetString(formatData, 0, formatData.Length);
                fieldNames = printer.GetFormatUtil().GetVariableFields(formatString);
            } catch (ZebraPrinterConnectionException) {
                updateGuiFromWorkerThread("Communication Error", Color.Red);
            }
            return fieldNames;
        }

        private void toggleMenuButtons(bool state) {
            rightSideButton.Enabled = state;
            leftSideButton.Enabled = state;

        }

        private void reprintButton_Click(object sender, EventArgs e) {
            SendLabelsToPrinter();
        }

        private void exitButton_Click(object sender, EventArgs e) {
            Close();
        }

        private void backToListFormatsButton_Click(object sender, EventArgs e) {
            ConnectedButtonSettings();
        }

        private void reEnterDataButton_Click(object sender, EventArgs e) {
            varFieldPanel.BringToFront();
        }

        private void StoredFormatDemoForm_Closing(object sender, CancelEventArgs e) {
            if (getConnection() != null) {
                getConnection().Close();
            }
            RemoveVariableFields();
        }

    }
}